alpenhorn/auto_import.py

```diff
@@ -67,54 +69,56 @@ def _import_file(node, root, acq_name, file_name):
         acq = ac.ArchiveAcq.get(ac.ArchiveAcq.name == acq_name)
         log.debug("Acquisition \"%s\" already in DB. Skipping." % acq_name)
     except pw.DoesNotExist:
-        acq = add_acq(acq_name)
+        acq = add_acq(acq_name, node)
         if acq is None:
             return
         log.info("Acquisition \"%s\" added to DB." % acq_name)
 
-    # TODO: file-type detection
     # What kind of file do we have?
-    ftype = detect_file_type(file_name)
+    ftype = ac.FileType.detect(file_name, acq, node)
+
     if ftype is None:
         log.info("Skipping unrecognised file \"%s/%s\"." % (acq_name, file_name))
         return
 
-    # TODO: acquisition-type specific handling
-
     # Add the file, if necessary.
     try:
         file = ac.ArchiveFile.get(ac.ArchiveFile.name == file_name,
                                   ac.ArchiveFile.acq == acq)
-        size_b = file.size_b
         log.debug("File \"%s/%s\" already in DB. Skipping." % (acq_name, file_name))
+
     except pw.DoesNotExist:
-        log.debug("Computing md5sum.")
+        log.debug("Computing md5sum of \"%s\"." % file_name)
         md5sum = md5sum_file(fullpath, cmd_line=False)
         size_b = os.path.getsize(fullpath)
+
         done = False
         while not done:
             try:
-                file = ac.ArchiveFile.create(acq=acq, type=ftype, name=file_name,
-                                             size_b=size_b, md5sum=md5sum)
+                with db.database_proxy.atomic():
+                    file_ = ac.ArchiveFile.create(acq=acq, type=ftype, name=file_name,
+                                                  size_b=size_b, md5sum=md5sum)
+
+                    ftype.file_info.new(file_, node)
+
                 done = True
-            except pw.OperationalError:
+            except pw.OperationalError as e:
+                log.exception(e)
                 log.error("MySQL connexion dropped. Will attempt to reconnect in "
                           "five seconds.")
                 time.sleep(5)
+
                 # di.connect_database(True)
         log.info("File \"%s/%s\" added to DB." % (acq_name, file_name))
 
     # Register the copy of the file here on the collection server, if (1) it does
     # not exist, or (2) it does exist but has been labelled as corrupt. If (2),
     # check again.
-    if not file.copies.where(ar.ArchiveFileCopy.node == node).count():
-        copy = ar.ArchiveFileCopy.create(file=file, node=node, has_file='Y',
+    if not file_.copies.where(ar.ArchiveFileCopy.node == node).count():
```

Comment:
`_file` is only set if the file is new, but this will fail if ArchiveFile does exist in the DB but does not have a corresponding ArchiveFileCopy on this node. We also don't check for the second case described as the comment above, if ArchiveFileCopy exists but has been marked as corrupt.